test/integration/garbagecollector/garbage_collector_test.go

```diff
@@ -710,3 +805,215 @@ func TestBlockingOwnerRefDoesBlock(t *testing.T) {
 		t.Errorf("expect there to be 1 pods, got %#v", pods.Items)
 	}
 }
+
+// TestCustomResourceCascadingDeletion ensures the basic cascading delete
+// behavior supports custom resources.
+func TestCustomResourceCascadingDeletion(t *testing.T) {
+	ctx := setup(t, 5)
+	defer ctx.tearDown()
+
+	clientSet, apiExtensionClient, clientPool := ctx.clientSet, ctx.apiExtensionClient, ctx.clientPool
+
+	ns := createNamespaceOrDie("crd-cascading", clientSet, t)
+
+	definition, resourceClient := createRandomCustomResourceDefinition(t, apiExtensionClient, clientPool, ns.Name)
+
+	// Create a custom owner resource.
+	owner := newCRDInstance(definition, ns.Name, names.SimpleNameGenerator.GenerateName("owner"))
+	owner, err := resourceClient.Create(owner)
+	if err != nil {
+		t.Fatalf("failed to create owner resource %q: %v", owner.GetName(), err)
+	}
+	t.Logf("created owner resource %q", owner.GetName())
+
+	// Create a custom dependent resource.
+	dependent := newCRDInstance(definition, ns.Name, names.SimpleNameGenerator.GenerateName("dependent"))
+	link(t, owner, dependent)
+
+	dependent, err = resourceClient.Create(dependent)
+	if err != nil {
+		t.Fatalf("failed to create dependent resource %q: %v", dependent.GetName(), err)
+	}
+	t.Logf("created dependent resource %q", dependent.GetName())
+
+	// Delete the owner.
+	foreground := metav1.DeletePropagationForeground
+	err = resourceClient.Delete(owner.GetName(), &metav1.DeleteOptions{PropagationPolicy: &foreground})
+	if err != nil {
+		t.Fatalf("failed to delete owner resource %q: %v", owner.GetName(), err)
+	}
+
+	// Ensure the owner is deleted.
+	if err := wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
+		_, err := resourceClient.Get(owner.GetName(), metav1.GetOptions{})
+		return errors.IsNotFound(err), nil
+	}); err != nil {
+		t.Fatalf("failed waiting for owner resource %q to be deleted", owner.GetName())
+	}
+
+	// Ensure the dependent is deleted.
+	_, err = resourceClient.Get(dependent.GetName(), metav1.GetOptions{})
+	if err == nil {
+		t.Fatalf("expected dependent %q to be deleted", dependent.GetName())
+	} else {
+		if !errors.IsNotFound(err) {
+			t.Fatalf("unexpected error getting dependent %q: %v", dependent.GetName(), err)
+		}
+	}
+}
+
+// TestMixedRelationships ensures that owner/dependent relationships work
+// between core and custom resources.
+//
+// TODO: Consider how this could be represented with table-style tests (e.g. a
+// before/after expected object graph given a delete operation targetting a
+// specific node in the before graph with certain delete options).
+func TestMixedRelationships(t *testing.T) {
```

Comment:
This is flaking frequently according to http://storage.googleapis.com/k8s-metrics/flakes-latest.json

See https://github.com/kubernetes/test-infra/tree/master/metrics